Src/WindowsServer/WindowsServer.Shared/Implementation/AssemblyLoader.cs

```diff
@@ -0,0 +1,90 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Reflection;
+
+    /// <summary>
+    /// AssemblyLoader contains logic to load a given assembly and read properties using reflection.
+    /// Inherits MarshalByRefObject so that methods of this class can be executed remotely in separate AppDomain.
+    /// </summary>    
+    internal class AssemblyLoader : MarshalByRefObject
+    {
+        internal string AssemblyName;
+        internal string Culture;
+        internal string PublicKeyToken;
+        internal string[] VersionsToAttempt;
+
+        public AssemblyLoader()
+        {
+            this.AssemblyName = "Microsoft.WindowsAzure.ServiceRuntime";
+            this.Culture = "neutral";
+            this.PublicKeyToken = "31bf3856ad364e35";
+            this.VersionsToAttempt = new string[] { "2.7.0.0", "2.6.0.0"};
+        }
+
+        public bool ReadAndPopulateContextInformation(ref string roleName, ref string roleInstanceId)
+        {
+            Assembly loadedAssembly = null;
+            try
+            {
+                // As this is executed inside a separate AppDomain, it is safe to load assemblies here without interfering with user code.                
+                loadedAssembly = this.AttemptToLoadAssembly(this.AssemblyName, this.Culture, this.PublicKeyToken, this.VersionsToAttempt);
+                if (loadedAssembly != null)
+                {
+                    ServiceRuntime serviceRuntime = new ServiceRuntime();
+                    RoleEnvironment roleEnvironment = serviceRuntime.GetRoleEnvironment(loadedAssembly);
+
+                    if (roleEnvironment.IsAvailable == true)
+                    {
+                        RoleInstance roleInstance = roleEnvironment.CurrentRoleInstance;
+                        if (roleInstance != null)
+                        {
+                            roleInstanceId = roleInstance.Id;
+                            Role role = roleInstance.Role;
+                            roleName = role.Name;
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                WindowsServerEventSource.Log.TroubleshootingMessageEvent("Unknown error occured attempting to populate Azure Context, Exception: " + ex.ToString());
+            }
+
+            return loadedAssembly != null;
+        }
+
+        private Assembly AttemptToLoadAssembly(string assemblyName, string culture, string publicKeyToken, string[] versionsToAttempt)
```

Comment:
I'd recommend again to use AssemblyName class. It knows how to format itself (you can pass the class itself to the `Assembly.Load`. One you'll switch to use of that class - this entire method will become redundant I believe